docs/architecture.md

```diff
@@ -0,0 +1,331 @@
+}
+
+trait Database {
+    // get current fuel blockchain height
+    fn get_current_height() -> Result<BlockHeight>;
+}
+```
+#### Adapters: fuel_core::service::adapters::fuel_sync
+_N/A_
+
+### P2P
+
+#### Ports: fuel_p2p::ports
+```rust
+trait Database {
+    // used when other peers are syncing
+    fn get_sealed_block(height: BlockHeight) -> Result<SealedFuelBlock>; 
+}
+```
+
+#### Adapters: fuel_core::service::adapters::peer_to_peer
+```rust
+impl fuel_poa_consensus::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+impl fuel_sync::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+impl fuel_txpool::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+```
+
+### Transaction Pool
+
+#### Ports: fuel_txpool::ports
+```rust
+trait PeerToPeer {
+    async fn broadcast_transaction(transaction: Transaction) -> Result<()>;
+    fn report_invalid_tx(peer_id: PeerId) -> Result<()>;
+    // handler for gossipped transactions
+    fn on_new_tx(tx: Transaction, peer: PeerId);
+}
+```
+
+#### Adapters: fuel_core::service::adapters::transaction_pool
+```rust
+impl block_importer::ports::TransactionPool for Service<TransactionPool> {
+    // insert impl here
+}
+
+impl block_producer::ports::TransactionPool for Service<TransactionPool> {
+    // insert impl here
+}
+```
+
+### Executor
+
+#### Ports: fuel_block_executor::ports
+```rust
+trait Database: IntepreterStorage {
+    fn get_coin(id: UtxoId) -> Result<Coin>;
+    fn upsert_coin(coin: Coin) -> Result<()>;
+    fn get_message(message_id: MessageId) -> Result<Message>;
+    fn mark_message_spent(messageId: MessageId) -> Result<()>;
+    fn get_contract_utxo_id(contract_id: ContractId) -> Result<UtxoId>;
+    fn set_contract_utxo_id(contract_id: ContractId, utxo_id: UtxoId) -> Result<()>;
+    fn get_block(height: BlockHeight) -> Result<FuelBlock>;
+    fn insert_block(block: FuelBlock) -> Result<()>;
+}
+```
+#### Adapters: fuel_core::service::adapters::executor
+```rust
+impl fuel_block_producer::ports::Executor for Service<Executor> {
+    // insert impl here
+}
+
+impl fuel_block_importer::ports::Executor for Service<Executor> {
+    // insert impl here
+}
+```
+
+### Relayer
+
+#### Ports: fuel_relayer::ports
+```rust
+trait Database {
+    fn get_current_sealed_block() -> Result<SealedFuelBlock>;
+    fn get_pending_relay_commitment() -> Option<PendingRelayCommit>;
+    fn set_pending_block_commitment(pending_commit: PendingRelayCommit) -> Result<()>;
+    fn clear_pending_block_commitment(pending_commit: PendingRelayCommit) -> Result<()>;
+    fn set_last_committed_fuel_height(height: BlockHeight) -> Result<()>;
+}
+```
+
+#### Adapters: fuel_core::service::adapters::relayer
+```rust
+impl fuel_block_importer::ports::Relayer for Service<Relayer> {
+    // insert impl here
+}
+
+impl fuel_block_producer::ports::Relayer for Service<Relayer> {
```

Comment:
Does fuel-block-producer need anything else specifying how blocks are produced? I was thinking the other day it might be nice to have some way for block production to be externally delegated to something over a public interface if made accessible. This greatly helps with a PBS or similar structure, however would then require changes in the public interface.

I think it could be very valuable to allow TxPool to delegate a `MakeBlock` function so that custom block building can easily integrate inside the txpool. Although we could also undertake this with a similar internal change of just having some `BlockProduction` Interface, and allowing people to swap in their own custom production methods via that.

I think it's worth weighing the 2 options, as a public interface allows more complex block building (especially if someone has tx's their txpool doesn't know about yet with a txpool aggregator across peergroups) and it also helps with something like PBS, but an inner strategy function avoids needing to have a public interface and is more simplistic.

Reply:
What is PBS?=D

Reply:
> Although we could also undertake this with a similar internal change of just having some BlockProduction Interface, and allowing people to swap in their own custom production methods via that.

This is the way to go. The consensus module defines it's own port for block production, and the adapter for it could use our defacto fuel_block_producer implementation or swap it out with something else if wanted.

If it's just a matter of overriding the selection algorithm of transactions, that could also be swapped out by the block producer's adapter for the txpool.

Reply:
> What is PBS?=D

PBS is Proposer Builder Seperation, so the idea that people like Flashbots and Manifold finance can build blocks for example that extract maximal MEV, then they can send these blocks to validators, who propose the blocks. Seperating Validators from Block Builders, it's the logical conclusion of Ethereum's MEV state, so I thought it would be worth considering in our new design, and allow for block building to be done separate from proposal